### PR TITLE
feat/jetpack_compose_integration

### PR DESCRIPTION
**Pull request by mattn02**
## Overview

This PR demos how to integrate the following ad formats using Kotlin's Jetpack Compose:

**AppLovin network ads:**
- Banner/Leader
- MREC

**MAX mediated ads:**
- Banner/Leader
- MREC
- Native ads using Templates API

## Testing

- Tested on local test device (Galaxy S20) using ad unit & application IDs from Test App.
- Used Profiler in Android Studios to ensure that no memory leaks occurred in new features.

### CODE REVIEW

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/interstitial/JetpackComposeInterstitialAdActivity.kt
Comment (Lorenzo45) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/interstitial/JetpackComposeInterstitialAdActivity.kt:1]: I actually misspoke in the Asana ticket description - we don't need Jetpack Compose integrations for _all_ of our ad formats. We really only need it for banners/leaders, MRECs, and native ads - we can skip fullscreen ad formats (interstitial, rewarded, app open).

Showing fullscreen ads just require API calls, but banners/MRECs/native ads are actually a part of the UI so Jetpack Compose demos make more sense there.

Could we remove all the fullscreen ad integrations to keep the PR clean and keep our demo app slimmer?

Looks like you got a lot of cool experience learning how to build Jetpack compose UI and display all our ad formats at least :)

File: AppLovin MAX Demo App - Kotlin/app/build.gradle [hunk 1]
```diff
@@ -48,7 +63,24 @@ dependencies {
...
+    implementation platform('androidx.compose:compose-bom:2022.10.00')
+    implementation 'androidx.navigation:navigation-fragment-ktx:2.6.0'
+    implementation 'androidx.navigation:navigation-ui-ktx:2.6.0'
+    implementation platform('androidx.compose:compose-bom:2022.10.00')
+    androidTestImplementation platform('androidx.compose:compose-bom:2022.10.00')
+    androidTestImplementation 'androidx.compose.ui:ui-test-junit4'
+    androidTestImplementation platform('androidx.compose:compose-bom:2022.10.00')
+    androidTestImplementation platform('androidx.compose:compose-bom:2022.10.00')
+    debugImplementation 'androidx.compose.ui:ui-tooling'
+    debugImplementation 'androidx.compose.ui:ui-test-manifest'
```
Comment (Lorenzo45) [inline AppLovin MAX Demo App - Kotlin/app/build.gradle:82]: You think we can remove any of these dependencies? We generally try to have as few dependencies as possible.

**Review by Lorenzo45 — CHANGES_REQUESTED** [review]

**Comment by Lorenzo45** [comment]
Also great to see you're already using the Android Studio profiler to check for memory leaks :)

**Comment by mattn02** [comment]
Changes made. Removed Jetpack Compose integration for fullscreen ads + cleaned up dependencies in gradle file

**Comment by mattn02** [comment]
Changes made: ensured Jetpack Compose features worked correctly with workspace environment & cleaned up changes in manifest to make PR slimmer.

I updated the packages for the MAX full screen ads (app open, rewarded, interstitial) after removing the jetpack compose integration since [the current one in master doesn't fully reflect the file location that they are in](https://github.com/AppLovin/AppLovin-MAX-SDK-Android/blob/dbc5fdedbb1479357945e6e3034d502a9951dbbc/AppLovin%20MAX%20Demo%20App%20-%20Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/AppOpenAdActivity.kt#L1). Let me know if I should keep these changes or revert it back.

**Review by avi-cramer — COMMENTED** [review]
Looks good, just a couple minor UI change requests:

- Could we get the "Load" button for AppLovin jetpack compose banners/mrecs to look more in line with the other activities? It is currently a little smaller and doesn't fit the color theme of the app.
- Indentation when we log the callbacks to match the other activities (for MAX and AppLovin activities)
- Jetpack Compose Native ads "Show Ad" button could use some margin on the bottom to match other activities
- Jetpack Compose Native ads are not displaying the same as normal Templates API, looks like the native ad is rendered in a box whereas for normal Templates API the ad is using the full screen width

**Comment by mattn02** [comment]
Changes made! Adjusted UI to match with rest of demo app + revamped architecture to follow MVVM convention as used in iOS PR :)

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/RewardedAdActivity.kt [hunk 2]
```diff
@@ -20,6 +20,7 @@ import kotlin.math.pow
  * <p>
  * Created by Harry Arakkal on 2019-09-17.
  */
+
```
Comment (avi-cramer) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/max/RewardedAdActivity.kt:23]:
```suggestion
```

**Review by avi-cramer — APPROVED** [review]
Looks good to me!

File: AppLovin MAX Demo App - Kotlin/app/build.gradle [hunk 3]
```diff
@@ -48,7 +58,12 @@ dependencies {
     implementation 'com.android.installreferrer:installreferrer:2.2'
 
     implementation 'com.google.android.gms:play-services-ads-identifier:17.1.0'
-    
+
+    implementation("androidx.compose.runtime:runtime:1.1.0")
+    implementation 'androidx.activity:activity-compose:1.3.0'
+    implementation "androidx.compose.material:material:1.3.0"
+    implementation 'androidx.compose.foundation:foundation:1.3.0'
```
Comment (Lorenzo45) [inline AppLovin MAX Demo App - Kotlin/app/build.gradle:65]: Thanks for slimming down the dependencies 👍 One small nit is that the `() " '` usages are inconsistent

File: AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/applovin/banners/BannerJetpackComposeActivity.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * Created by Matthew Nguyen on 2023-07-26.
+ */
+class BannerJetpackComposeActivity : BaseJetpackComposeAdActivity() {
+    private lateinit var viewModel: AppLovinAdViewComposableViewModel
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // Initialize view model.
+        viewModel = AppLovinAdViewComposableViewModel(this)
```
Comment (Lorenzo45) [inline AppLovin MAX Demo App - Kotlin/app/src/main/java/com/applovin/enterprise/apps/demoapp/ads/applovin/banners/BannerJetpackComposeActivity.kt:30]: nit: `viewModel` is pretty generic, would be good to specify that this is an `bannerViewModel` or something (same comment applies to the other Activities)

**Review by Lorenzo45 — APPROVED** [review]
Looks super clean overall! I think the view model architecture is great. Nice work :)

**Comment by mattn02** [comment]
Changes made!

**Review by thomasmso — APPROVED** [review]

**Pull request merged**